byway/src/main/java/com/google/sps/data/Trip.java

```diff
@@ -0,0 +1,88 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.data;
+
+import com.google.appengine.api.datastore.Entity;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import java.util.Collection;
+import java.util.ArrayList;
+
+/**
+ * A class to make a Trip type, containing
+ * a specific id, destinations, interests and routes 
+ * to be used when updating a trip.
+ */
+public final class Trip {
+
+  private final String id;
+  private final String start;
+  private final ArrayList<String> destinations;
+  private final ArrayList<String> interests;
+  private final ArrayList<String> route;
+  public static final String DATASTORE_ENTITY_KIND = "Trip";
+
+  public Trip(String id, String start, Collection<String> destinations,
+              Collection<String> interests, Collection<String> route) {
+    this.id = checkNotNull(id, "id");
+    this.start = checkNotNull(start, "start");
+    Collection<String> validDestinations = checkNotNull(destinations, "destinations");
+    Collection<String> validInterests = checkNotNull(interests, "interests");
+    Collection<String> validRoute = checkNotNull(route, "route");
+    this.destinations = new ArrayList<String>(validDestinations);
+    this.interests = new ArrayList<String>(validInterests);
+    this.route = new ArrayList<String>(validRoute);
+  }
+
+  public ArrayList<String> getInterests() {
+    return new ArrayList<String>(this.interests);
```

Comment:
Not sure if I understand. How could the inner list change and this unmodifiable list would show the updated values, since it's unmodifiable and changing the inner list means modifying it? Maybe I'm misinterpreting which list is the inner list, Is the inner list the list within the class? If so, we wouldn't ever be able to modify (as of now with no setters) unless we create a new Trip, in which case it is trivial. If the inner list is the list created before the Trip class is made, I'm assuming the whole point of having the unmodifiable list this way means the changed won't be made inside the class, even if changed after it were declared.

Reply:
`unmodifiableList` returns a new `List` object with a reference to the original list inside it. When you do any non-modifying operation like `get` it just calls `get` on the inner list. If you try to call any method that would modify the original list, like `add`, it throws an exception. It's still possible to call `add` on the arraylist itself, but the wrapper list prevents modification through that wrapper.

The code for this is public and not actually that hard to read: https://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/java/util/Collections.java#l1355

Notice how it keeps a reference to the original list: `final List<? extends E> list;`, and look at the implementations for `get` and `set`:

```java
        public E get(int index) {return list.get(index);} // delegates this operation to the inner list.
        public E set(int index, E element) { // rejects if you try to set anything.
            throw new UnsupportedOperationException();
        }
```

Since this is just a wrapper object around another list, the inner list is still perfectly usable and still supports operations that add things to it, as long as you call `add` on a reference to the original list that you've saved elsewhere, like `this.interests`, but will not allow additions or changes through the wrapper class.

Reply:
I see, makes more sense - thanks